File: internal/metrics/metrics.go
// Package metrics provides Prometheus metrics collection and management
// for the TLS Certificate Monitor. It tracks certificate health, security
// issues, and operational statistics.
package metrics

import (
	"errors"
	"fmt"
	"sync"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/collectors"
	dto "github.com/prometheus/client_model/go"
)

var (
	once     sync.Once
	instance *Collector

	// Global flag to track if runtime metrics have been registered
	// This prevents duplicate registrations across all registries
	runtimeMetricsRegistered bool
	runtimeMetricsMutex      sync.Mutex
)

// Collector manages all Prometheus metrics
// Field order optimized for memory alignment (fieldalignment fix)
type Collector struct {
	// Gauge vectors (8 bytes each pointer)
	certExpiration     *prometheus.GaugeVec
	certSANCount       *prometheus.GaugeVec
	certInfo           *prometheus.GaugeVec
	certDuplicateCount *prometheus.GaugeVec
	certIssuerCode     *prometheus.GaugeVec

	// Simple gauges (8 bytes each)
	weakKeyTotal         prometheus.Gauge
	deprecatedSigAlg     prometheus.Gauge
	certFilesTotal       prometheus.Gauge
	certsParsedTotal     prometheus.Gauge
	certParseErrorsTotal prometheus.Gauge
	scanDuration         prometheus.Gauge
	lastScanTimestamp    prometheus.Gauge

	// Other fields
	registry prometheus.Registerer // 16 bytes (interface)
	mu       sync.RWMutex          // 24 bytes
}

// NewCollector creates a new metrics collector (singleton for default registry)
func NewCollector() *Collector {
	once.Do(func() {
		instance = createCollector(prometheus.DefaultRegisterer)
	})
	return instance
}

// NewCollectorWithRegistry creates a new metrics collector with a custom registry (for testing)
func NewCollectorWithRegistry(reg prometheus.Registerer) *Collector {
	return createCollector(reg)
}

// createCollector creates the actual collector instance
func createCollector(reg prometheus.Registerer) *Collector {
	c := &Collector{
		registry: reg,
		// Certificate metrics
		certExpiration: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "ssl_cert_expiration_timestamp",
				Help: "Certificate expiration time (Unix timestamp)",
			},
			[]string{"path", "subject", "issuer"},
		),
		certSANCount: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "ssl_cert_san_count",
				Help: "Number of Subject Alternative Names",
			},
			[]string{"path"},
		),
		certInfo: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "ssl_cert_info",
				Help: "Certificate information with labels",
			},
			[]string{"path", "subject", "issuer", "serial", "signature_algorithm"},
		),
		certDuplicateCount: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "ssl_cert_duplicate_count",
				Help: "Number of duplicate certificates",
			},
			[]string{"fingerprint"},
		),
		certIssuerCode: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "ssl_cert_issuer_code",
				Help: "Numeric issuer classification",
			},
			[]string{"issuer", "common_name", "file_name"},
		),

		// Security metrics
		weakKeyTotal: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_cert_weak_key_total",
				Help: "Certificates with weak cryptographic keys",
			},
		),
		deprecatedSigAlg: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_cert_deprecated_sigalg_total",
				Help: "Certificates using deprecated signature algorithms",
			},
		),

		// Operational metrics
		certFilesTotal: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_cert_files_total",
				Help: "Total certificate files processed",
			},
		),
		certsParsedTotal: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_certs_parsed_total",
				Help: "Successfully parsed certificates",
			},
		),
		certParseErrorsTotal: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_cert_parse_errors_total",
				Help: "Certificate parsing errors",
			},
		),
		scanDuration: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_cert_scan_duration_seconds",
				Help: "Directory scan duration",
			},
		),
		lastScanTimestamp: prometheus.NewGauge(
			prometheus.GaugeOpts{
				Name: "ssl_cert_last_scan_timestamp",
				Help: "Last successful scan time",
			},
		),
	}

	// Register all metrics with the provided registerer
	c.registerMetrics(reg)

	return c
}

// safeRegister safely registers a collector, logging warnings instead of panicking on duplicates
func (c *Collector) safeRegister(reg prometheus.Registerer, collector prometheus.Collector, name string) {
	if err := reg.Register(collector); err != nil {
		// Use errors.As for proper error checking (errorlint fix)
		var alreadyRegisteredError prometheus.AlreadyRegisteredError
		if !errors.As(err, &alreadyRegisteredError) {
			// Only log error for non-duplicate registration issues
			fmt.Printf("Warning: Failed to register metric %s: %v\n", name, err)
		}
		// For duplicate registrations, silently continue as this is expected during testing
	}
}

// registerMetrics registers all metrics with the provided registerer
func (c *Collector) registerMetrics(reg prometheus.Registerer) {
	// Certificate metrics - use safe registration
	c.safeRegister(reg, c.certExpiration, "ssl_cert_expiration_timestamp")
	c.safeRegister(reg, c.certSANCount, "ssl_cert_san_count")
	c.safeRegister(reg, c.certInfo, "ssl_cert_info")
	c.safeRegister(reg, c.certDuplicateCount, "ssl_cert_duplicate_count")
	c.safeRegister(reg, c.certIssuerCode, "ssl_cert_issuer_code")

	// Security metrics
	c.safeRegister(reg, c.weakKeyTotal, "ssl_cert_weak_key_total")
	c.safeRegister(reg, c.deprecatedSigAlg, "ssl_cert_deprecated_sigalg_total")

	// Operational metrics
	c.safeRegister(reg, c.certFilesTotal, "ssl_cert_files_total")
	c.safeRegister(reg, c.certsParsedTotal, "ssl_certs_parsed_total")
	c.safeRegister(reg, c.certParseErrorsTotal, "ssl_cert_parse_errors_total")
	c.safeRegister(reg, c.scanDuration, "ssl_cert_scan_duration_seconds")
	c.safeRegister(reg, c.lastScanTimestamp, "ssl_cert_last_scan_timestamp")

	// Only register Go runtime metrics for the default registry and only once globally
	c.registerRuntimeMetricsIfNeeded(reg)
}

// registerRuntimeMetricsIfNeeded registers Go runtime metrics only once globally
// and only for the default registry to avoid conflicts
func (c *Collector) registerRuntimeMetricsIfNeeded(reg prometheus.Registerer) {
	// Only proceed if this is the default registry
	if reg != prometheus.DefaultRegisterer {
		return
	}

	runtimeMetricsMutex.Lock()
	defer runtimeMetricsMutex.Unlock()

	// Check if runtime metrics have already been registered
	if runtimeMetricsRegistered {
		return
	}

	// Try to register runtime metrics
	goCollector := collectors.NewGoCollector()
	processCollector := collectors.NewProcessCollector(collectors.ProcessCollectorOpts{})

	// Use the safe register method to avoid panics
	c.safeRegister(reg, goCollector, "go_collector")
	c.safeRegister(reg, processCollector, "process_collector")

	// Mark as registered regardless of success/failure to prevent retry spam
	runtimeMetricsRegistered = true
}

// ResetCertificateMetrics resets certificate-specific metrics
func (c *Collector) ResetCertificateMetrics() {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.certExpiration.Reset()
	c.certSANCount.Reset()
	c.certInfo.Reset()
	c.certDuplicateCount.Reset()
	c.certIssuerCode.Reset()
}

// SetCertExpiration sets certificate expiration metric
func (c *Collector) SetCertExpiration(path, subject, issuer string, timestamp float64) {
	c.certExpiration.WithLabelValues(path, subject, issuer).Set(timestamp)
}

// SetCertSANCount sets SAN count metric
func (c *Collector) SetCertSANCount(path string, count float64) {
	c.certSANCount.WithLabelValues(path).Set(count)
}

// SetCertInfo sets certificate info metric
func (c *Collector) SetCertInfo(path, subject, issuer, serial, sigAlg string) {
	c.certInfo.WithLabelValues(path, subject, issuer, serial, sigAlg).Set(1)
}

// SetCertDuplicateCount sets duplicate count metric
func (c *Collector) SetCertDuplicateCount(fingerprint string, count float64) {
	c.certDuplicateCount.WithLabelValues(fingerprint).Set(count)
}

// SetCertIssuerCode sets issuer code metric (legacy method for backward compatibility)
func (c *Collector) SetCertIssuerCode(issuer string, code float64) {
	c.certIssuerCode.WithLabelValues(issuer, "", "").Set(code)
}

// SetCertIssuerCodeWithLabels sets issuer code metric with additional labels
func (c *Collector) SetCertIssuerCodeWithLabels(issuer, commonName, fileName string, code float64) {
	c.certIssuerCode.WithLabelValues(issuer, commonName, fileName).Set(code)
}

// SetWeakKeyTotal sets weak key total metric
func (c *Collector) SetWeakKeyTotal(total float64) {
	c.weakKeyTotal.Set(total)
}

// SetDeprecatedSigAlgTotal sets deprecated signature algorithm total metric
func (c *Collector) SetDeprecatedSigAlgTotal(total float64) {
	c.deprecatedSigAlg.Set(total)
}

// SetCertFilesTotal sets total certificate files metric
func (c *Collector) SetCertFilesTotal(total float64) {
	c.certFilesTotal.Set(total)
}

// SetCertsParsedTotal sets parsed certificates total metric
func (c *Collector) SetCertsParsedTotal(total float64) {
	c.certsParsedTotal.Set(total)
}

// SetCertParseErrorsTotal sets parse errors total metric
func (c *Collector) SetCertParseErrorsTotal(total float64) {
	c.certParseErrorsTotal.Set(total)
}

// SetScanDuration sets scan duration metric
func (c *Collector) SetScanDuration(seconds float64) {
	c.scanDuration.Set(seconds)
}

// SetLastScanTimestamp sets last scan timestamp metric
func (c *Collector) SetLastScanTimestamp(timestamp float64) {
	c.lastScanTimestamp.Set(timestamp)
}

// GetMetrics returns current metric values for health checks
func (c *Collector) GetMetrics() map[string]float64 {
	c.mu.RLock()
	defer c.mu.RUnlock()

	metrics := make(map[string]float64)

	// Gather current values
	metrics["cert_files_total"] = c.getGaugeValue(c.certFilesTotal)
	metrics["certs_parsed_total"] = c.getGaugeValue(c.certsParsedTotal)
	metrics["cert_parse_errors_total"] = c.getGaugeValue(c.certParseErrorsTotal)
	metrics["weak_key_total"] = c.getGaugeValue(c.weakKeyTotal)
	metrics["deprecated_sigalg_total"] = c.getGaugeValue(c.deprecatedSigAlg)
	metrics["last_scan_timestamp"] = c.getGaugeValue(c.lastScanTimestamp)

	return metrics
}

// getGaugeValue safely retrieves a gauge value
func (c *Collector) getGaugeValue(gauge prometheus.Gauge) float64 {
	metric := &dto.Metric{}
	// Handle error from Write (errcheck fix)
	if err := gauge.Write(metric); err != nil {
		// Log error but return zero value
		fmt.Printf("Failed to read gauge value: %v\n", err)
		return 0
	}
	if metric.Gauge != nil && metric.Gauge.Value != nil {
		return *metric.Gauge.Value
	}
	return 0
}

File: tls_cert_monitor/metrics.py
"""
Prometheus metrics collection for TLS Certificate Monitor.
"""

import socket
import time
from collections import defaultdict
from typing import Any, Dict, List

import psutil
from prometheus_client import (
    CONTENT_TYPE_LATEST,
    CollectorRegistry,
    Gauge,
    Histogram,
    Info,
    generate_latest,
)

from tls_cert_monitor.logger import get_logger, log_metrics_collection


class MetricsCollector:
    """Prometheus metrics collector for TLS certificates and application metrics."""

    def __init__(self) -> None:
        self.logger = get_logger("metrics")
        self.registry = CollectorRegistry()

        # Certificate metrics
        self.ssl_cert_expiration_timestamp = Gauge(
            "ssl_cert_expiration_timestamp",
            "Certificate expiration time (Unix timestamp)",
            ["common_name", "issuer", "path", "serial"],
            registry=self.registry,
        )

        self.ssl_cert_san_count = Gauge(
            "ssl_cert_san_count",
            "Number of Subject Alternative Names",
            ["common_name", "path"],
            registry=self.registry,
        )

        self.ssl_cert_info = Info(
            "ssl_cert_info",
            "Certificate information with labels",
            ["path", "common_name", "issuer", "serial", "subject"],
            registry=self.registry,
        )

        self.ssl_cert_duplicate_count = Gauge(
            "ssl_cert_duplicate_count", "Number of duplicate certificates", registry=self.registry
        )

        # Fixed: Simplified duplicate names metric to avoid label conflicts
        self.ssl_cert_duplicate_names = Info(
            "ssl_cert_duplicate_names",
            "Names of certificates that are duplicates",
            ["serial_number"],  # Changed from 'serial' and removed 'paths'
            registry=self.registry,
        )

        self.ssl_cert_issuer_code = Gauge(
            "ssl_cert_issuer_code",
            "Numeric issuer classification",
            ["common_name", "issuer", "path"],
            registry=self.registry,
        )

        # Cryptographic security metrics
        self.ssl_cert_weak_key_total = Gauge(
            "ssl_cert_weak_key_total",
            "Current count of certificates with weak cryptographic keys",
            registry=self.registry,
        )

        self.ssl_cert_deprecated_sigalg_total = Gauge(
            "ssl_cert_deprecated_sigalg_total",
            "Current count of certificates using deprecated signature algorithms",
            registry=self.registry,
        )

        # Operational metrics
        self.ssl_cert_files_total = Gauge(
            "ssl_cert_files_total",
            "Total certificate files processed",
            ["directory"],
            registry=self.registry,
        )

        self.ssl_certs_parsed_total = Gauge(
            "ssl_certs_parsed_total", "Successfully parsed certificates", registry=self.registry
        )

        self.ssl_cert_parse_errors_total = Gauge(
            "ssl_cert_parse_errors_total",
            "Current count of certificate parsing errors",
            registry=self.registry,
        )

        self.ssl_cert_parse_error_names = Info(
            "ssl_cert_parse_error_names",
            "Names of certificates that have parsing errors",
            ["filename", "error_type", "error_message"],
            registry=self.registry,
        )

        self.ssl_cert_scan_duration_seconds = Histogram(
            "ssl_cert_scan_duration_seconds",
            "Directory scan duration",
            ["directory"],
            registry=self.registry,
        )

        self.ssl_cert_last_scan_timestamp = Gauge(
            "ssl_cert_last_scan_timestamp",
            "Last successful scan time",
            ["directory"],
            registry=self.registry,
        )

        # Application metrics
        self.app_memory_bytes = Gauge(
            "app_memory_bytes",
            "Application memory usage in bytes",
            ["type"],  # rss, vms, shared, etc.
            registry=self.registry,
        )

        self.app_cpu_percent = Gauge(
            "app_cpu_percent", "Application CPU usage percentage", registry=self.registry
        )

        self.app_thread_count = Gauge(
            "app_thread_count", "Number of application threads", registry=self.registry
        )

        self.app_info = Info(
            "app_info",
            "Application information",
            ["hostname", "version", "python_version"],
            registry=self.registry,
        )

        # Internal tracking
        self._duplicate_certificates: Dict[str, List[str]] = defaultdict(list)
        self._current_scan_parse_errors = 0  # Count of parse errors in current scan
        self._current_scan_weak_keys = 0  # Count of weak keys in current scan
        self._current_scan_deprecated_sigalgs = (
            0  # Count of deprecated signature algorithms in current scan
        )
        self._last_system_update = 0.0
        self._system_update_interval = 30  # Update system metrics every 30 seconds

        self.logger.info("Metrics collector initialized")

    def update_certificate_metrics(self, cert_data: Dict[str, Any]) -> None:
        """
        Update certificate-related metrics.

        Args:
            cert_data: Certificate data dictionary
        """
        try:
            common_name = cert_data.get("common_name", "unknown")
            issuer = cert_data.get("issuer", "unknown")
            path = cert_data.get("path", "unknown")
            serial = cert_data.get("serial", "unknown")

            # Expiration timestamp
            if "expiration_timestamp" in cert_data:
                self.ssl_cert_expiration_timestamp.labels(
                    common_name=common_name, issuer=issuer, path=path, serial=serial
                ).set(float(cert_data["expiration_timestamp"]))

            # SAN count
            if "san_count" in cert_data:
                self.ssl_cert_san_count.labels(common_name=common_name, path=path).set(
                    int(cert_data["san_count"])
                )

            # Certificate info
            self.ssl_cert_info.labels(
                path=path,
                common_name=common_name,
                issuer=issuer,
                serial=serial,
                subject=cert_data.get("subject", "unknown"),
            ).info({})

            # Issuer code
            issuer_code = self._get_issuer_code(issuer)
            self.ssl_cert_issuer_code.labels(common_name=common_name, issuer=issuer, path=path).set(
                issuer_code
            )

            # Track duplicates by serial number
            if serial != "unknown":
                self._duplicate_certificates[serial].append(path)

            # Check for weak keys
            if cert_data.get("is_weak_key", False):
                self._current_scan_weak_keys += 1

            # Check for deprecated signature algorithms
            if cert_data.get("is_deprecated_algorithm", False):
                self._current_scan_deprecated_sigalgs += 1

            log_metrics_collection(
                self.logger,
                "certificate_processed",
                1.0,
                {"common_name": common_name, "path": path},
            )

        except Exception as e:
            self.logger.error(f"Failed to update certificate metrics: {e}")

    def update_scan_metrics(
        self,
        directory: str,
        duration: float,
        files_total: int,
        parsed_total: int,
        errors_total: int,
    ) -> None:
        """
        Update scan-related metrics.

        Args:
            directory: Scanned directory
            duration: Scan duration in seconds
            files_total: Total files processed
            parsed_total: Successfully parsed certificates
            errors_total: Number of parsing errors
        """
        try:
            self.ssl_cert_files_total.labels(directory=directory).set(int(files_total))
            self.ssl_cert_scan_duration_seconds.labels(directory=directory).observe(duration)
            self.ssl_cert_last_scan_timestamp.labels(directory=directory).set(int(time.time()))

            # Set current counts (not cumulative)
            self.ssl_certs_parsed_total.set(parsed_total)
            self.ssl_cert_parse_errors_total.set(self._current_scan_parse_errors)
            self.ssl_cert_weak_key_total.set(self._current_scan_weak_keys)
            self.ssl_cert_deprecated_sigalg_total.set(self._current_scan_deprecated_sigalgs)

            log_metrics_collection(
                self.logger,
                "scan_completed",
                duration,
                {
                    "directory": directory,
                    "files_total": files_total,
                    "parsed_total": parsed_total,
                    "errors_total": errors_total,
                },
            )

        except Exception as e:
            self.logger.error(f"Failed to update scan metrics: {e}")

    def record_parse_error(self, filename: str, error_type: str, error_message: str) -> None:
        """
        Record a certificate parsing error.

        Args:
            filename: Name of the file that failed to parse
            error_type: Type of error
            error_message: Error message
        """
        try:
            # Increment our internal counter for current scan
            self._current_scan_parse_errors += 1

            self.ssl_cert_parse_error_names.labels(
                filename=filename,
                error_type=error_type,
                error_message=error_message[:100],  # Truncate long messages
            ).info({"full_error": error_message, "timestamp": str(time.time())})

            log_metrics_collection(
                self.logger, "parse_error", 1.0, {"filename": filename, "error_type": error_type}
            )

        except Exception as e:
            self.logger.error(f"Failed to record parse error: {e}")

    def update_duplicate_metrics(self) -> None:
        """Update duplicate certificate metrics."""
        try:
            # Count duplicates (serials with more than one path)
            duplicates = {
                serial: paths
                for serial, paths in self._duplicate_certificates.items()
                if len(paths) > 1
            }

            self.ssl_cert_duplicate_count.set(int(len(duplicates)))

            # Record duplicate information with fixed labels
            for serial, paths in duplicates.items():
                self.ssl_cert_duplicate_names.labels(serial_number=serial).info(
                    {
                        "duplicate_count": str(len(paths)),
                        "certificate_paths": ",".join(paths),
                        "path_count": str(len(paths)),
                    }
                )

            if duplicates:
                log_metrics_collection(self.logger, "duplicates_found", len(duplicates))

        except Exception as e:
            self.logger.error(f"Failed to update duplicate metrics: {e}")

    def update_system_metrics(self) -> None:
        """Update system and application metrics."""
        current_time = time.time()

        # Only update system metrics every N seconds to reduce overhead
        if current_time - self._last_system_update < self._system_update_interval:
            return

        try:
            # Get current process
            process = psutil.Process()

            # Memory metrics
            memory_info = process.memory_info()
            self.app_memory_bytes.labels(type="rss").set(int(memory_info.rss))
            self.app_memory_bytes.labels(type="vms").set(int(memory_info.vms))

            # CPU metrics
            cpu_percent = process.cpu_percent()
            self.app_cpu_percent.set(cpu_percent)

            # Thread count
            thread_count = process.num_threads()
            self.app_thread_count.set(int(thread_count))

            # Application info (only set once)
            import sys

            from tls_cert_monitor import __version__

            self.app_info.labels(
                hostname=socket.gethostname(),
                version=__version__,
                python_version=f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
            ).info({"platform": sys.platform, "process_id": str(process.pid)})

            self._last_system_update = current_time

            log_metrics_collection(
                self.logger,
                "system_metrics_updated",
                1.0,
                {
                    "memory_rss": memory_info.rss,
                    "cpu_percent": cpu_percent,
                    "thread_count": thread_count,
                },
            )

        except Exception as e:
            self.logger.error(f"Failed to update system metrics: {e}")

    def reset_scan_metrics(self) -> None:
        """Reset scan-specific metrics for a new scan. Resets current counts but preserves historical data."""
        self._duplicate_certificates.clear()
        self._current_scan_parse_errors = 0
        self._current_scan_weak_keys = 0
        self._current_scan_deprecated_sigalgs = 0

        # Immediately reset the gauge metrics to zero for instant feedback
        self.ssl_certs_parsed_total.set(0)
        self.ssl_cert_parse_errors_total.set(0)
        self.ssl_cert_weak_key_total.set(0)
        self.ssl_cert_deprecated_sigalg_total.set(0)
        self.ssl_cert_duplicate_count.set(0)

        self.logger.debug("Scan metrics reset (current counts cleared and gauges zeroed)")

    def clear_all_certificate_metrics(self) -> None:
        """Clear all labeled certificate metrics. Used when exclude patterns change."""
        try:
            # Clear all labeled metrics by recreating them
            # This is necessary because Prometheus labeled metrics can't be selectively cleared

            # Store the registry reference
            registry = self.registry

            # Unregister all labeled certificate metrics
            labeled_metrics = [
                self.ssl_cert_expiration_timestamp,
                self.ssl_cert_san_count,
                self.ssl_cert_info,
                self.ssl_cert_duplicate_names,
                self.ssl_cert_issuer_code,
            ]

            for metric in labeled_metrics:
                try:
                    registry.unregister(metric)
                except KeyError:
                    # Metric might not be registered yet
                    pass

            # Recreate all labeled metrics
            self.ssl_cert_expiration_timestamp = Gauge(
                "ssl_cert_expiration_timestamp",
                "Certificate expiration time (Unix timestamp)",
                ["common_name", "issuer", "path", "serial"],
                registry=registry,
            )

            self.ssl_cert_san_count = Gauge(
                "ssl_cert_san_count",
                "Number of Subject Alternative Names",
                ["common_name", "path"],
                registry=registry,
            )

            self.ssl_cert_info = Info(
                "ssl_cert_info",
                "Certificate information with labels",
                ["path", "common_name", "issuer", "serial", "subject"],
                registry=registry,
            )

            self.ssl_cert_duplicate_names = Info(
                "ssl_cert_duplicate_names",
                "Names of certificates that are duplicates",
                ["serial_number"],
                registry=registry,
            )

            self.ssl_cert_issuer_code = Gauge(
                "ssl_cert_issuer_code",
                "Numeric issuer classification",
                ["common_name", "issuer", "path"],
                registry=registry,
            )

            self.logger.debug("All labeled certificate metrics cleared and recreated")

        except Exception as e:
            self.logger.error(f"Failed to clear certificate metrics: {e}")

    def reset_parse_error_metrics(self) -> None:
        """Reset parse error metrics - useful after configuration changes like new passwords."""
        # Reset the current scan error count and gauge
        self._current_scan_parse_errors = 0
        self.ssl_cert_parse_errors_total.set(0)

        # Clear individual error details
        try:
            self.registry.unregister(self.ssl_cert_parse_error_names)
        except KeyError:
            # Metric might not be registered yet
            pass

        self.ssl_cert_parse_error_names = Info(
            "ssl_cert_parse_error_names",
            "Names of certificates that have parsing errors",
            ["filename", "error_type", "error_message"],
            registry=self.registry,
        )

        self.logger.debug("Parse error metrics recreated")

    def get_metrics(self) -> str:
        """
        Get Prometheus metrics in text format.

        Returns:
            Metrics in Prometheus text format
        """
        # Update system metrics before generating output
        self.update_system_metrics()

        # Update duplicate metrics
        self.update_duplicate_metrics()

        # Get raw metrics
        raw_metrics = generate_latest(self.registry).decode("utf-8")

        # Format numeric values to remove scientific notation and unnecessary decimals
        formatted_metrics = self._format_numeric_values(raw_metrics)

        return formatted_metrics

    def _format_numeric_values(self, metrics_text: str) -> str:
        """
        Format numeric values in metrics to use integers where appropriate.

        Args:
            metrics_text: Raw Prometheus metrics text

        Returns:
            Formatted metrics text with integers instead of scientific notation/decimals
        """
        import re

        lines = metrics_text.split("\n")
        formatted_lines = []

        for line in lines:
            if line.startswith("#") or not line.strip():
                # Keep comments and empty lines unchanged
                formatted_lines.append(line)
                continue

            # Match lines with metric values (with or without labels)
            match = re.match(r"^([^}]+})\s+(.+)$", line) or re.match(r"^([^\s]+)\s+(.+)$", line)
            if match:
                metric_name = match.group(1)
                value = match.group(2)

                # Convert scientific notation and decimals to integers for specific metrics
                if any(
                    metric in metric_name
                    for metric in [
                        "ssl_cert_last_scan_timestamp",
                        "ssl_cert_san_count",
                        "ssl_cert_files_total",
                        "ssl_cert_duplicate_count",
                        "app_memory_bytes",
                        "app_thread_count",
                        "ssl_cert_issuer_code",
                    ]
                ):
                    try:
                        # Convert scientific notation and float to integer
                        if "e+" in value:
                            int_value = int(float(value))
                        elif value.endswith(".0"):
                            int_value = int(float(value))
                        else:
                            # Try to convert, fallback to original if not numeric
                            try:
                                float_value = float(value)
                                if float_value.is_integer():
                                    int_value = int(float_value)
                                else:
                                    int_value = None
                            except ValueError:
                                int_value = None

                        if int_value is not None:
                            formatted_lines.append(f"{metric_name} {int_value}")
                        else:
                            formatted_lines.append(line)
                    except (ValueError, OverflowError):
                        # If conversion fails, keep original
                        formatted_lines.append(line)
                else:
                    # Keep other metrics unchanged (like CPU percentages)
                    formatted_lines.append(line)
            else:
                formatted_lines.append(line)

        return "\n".join(formatted_lines)

    def get_content_type(self) -> str:
        """Get content type for metrics endpoint."""
        return CONTENT_TYPE_LATEST

    def _get_issuer_code(self, issuer: str) -> int:
        """
        Get numeric issuer classification.

        Args:
            issuer: Certificate issuer string

        Returns:
            Numeric issuer code
        """
        issuer_lower = issuer.lower()

        # DigiCert
        if "digicert" in issuer_lower:
            return 30

        # Amazon
        if any(keyword in issuer_lower for keyword in ["amazon", "aws"]):
            return 31

        # Self-signed
        if any(keyword in issuer_lower for keyword in ["self-signed", "localhost", "127.0.0.1"]):
            return 33

        # Other/Unknown
        return 32

    def get_registry_status(self) -> Dict[str, Any]:
        """Get Prometheus registry status for health checks."""
        try:
            # Count metrics by type
            metrics_count = len(list(self.registry._collector_to_names.keys()))

            return {
                "prometheus_registry": {
                    "status": "healthy",
                    "metrics_count": metrics_count,
                    "last_update": self._last_system_update,
                }
            }
        except Exception as e:
            return {"prometheus_registry": {"status": "error", "error": str(e)}}


# Utility functions for metric helpers
def is_weak_key(key_size: int, algorithm: str) -> bool:
    """
    Check if a key is considered weak.

    Args:
        key_size: Key size in bits
        algorithm: Key algorithm

    Returns:
        True if key is weak
    """
    algorithm_lower = algorithm.lower()

    # Check ECDSA first (before RSA) since "ecdsa" contains "rsa"
    if "ec" in algorithm_lower or "ecdsa" in algorithm_lower:
        # ECDSA: P-256 (256 bits) is considered secure, anything below is weak
        return key_size < 256
    elif "rsa" in algorithm_lower:
        return key_size < 2048
    elif "dsa" in algorithm_lower:
        return key_size < 2048

    # Unknown algorithm, be conservative
    return key_size < 2048


def is_deprecated_signature_algorithm(algorithm: str) -> bool:
    """
    Check if a signature algorithm is deprecated.

    Args:
        algorithm: Signature algorithm

    Returns:
        True if algorithm is deprecated
    """
    algorithm_lower = algorithm.lower()

    deprecated_algorithms = ["md5", "sha1", "md2", "md4"]

    return any(alg in algorithm_lower for alg in deprecated_algorithms)
